### PR TITLE
Always CASCADE while dropping a shard

### LINKED ISSUE #927: Dropping a table which is referenced via foreign key does not work as expected
Steps to re-produce:

``` SQL

SET citus.shard_count = 2;
SET citus.shard_replication_factor = 1;

CREATE TABLE raw_events_first (user_id int, time timestamp, value_1 int, value_2 int, value_3 float, value_4 bigint, UNIQUE(user_id));
SELECT create_distributed_table('raw_events_first', 'user_id');

CREATE TABLE raw_events_second (user_id int REFERENCES raw_events_first(user_id), time timestamp, value_1 int, value_2 int, value_3 float, value_4 bigint, UNIQUE(user_id));
SELECT create_distributed_table('raw_events_second', 'user_id');

DROP TABLE raw_events_first CASCADE;


NOTICE:  drop cascades to constraint raw_events_second_user_id_fkey on table raw_events_second
WARNING:  cannot drop table raw_events_first_102008 because other objects depend on it
DETAIL:  constraint raw_events_second_user_id_fkey_102010 on table raw_events_second_102010 depends on table raw_events_first_102008
HINT:  Use DROP ... CASCADE to drop the dependent objects too.
CONTEXT:  while executing command on localhost:9700
SQL statement "SELECT master_drop_all_shards(v_obj.objid, v_obj.schema_name, v_obj.object_name)"
PL/pgSQL function citus_drop_trigger() line 40 at PERFORM
WARNING:  could not delete shard "raw_events_first_102008" on node "localhost:9700"
DETAIL:  Marking this shard placement for deletion
CONTEXT:  SQL statement "SELECT master_drop_all_shards(v_obj.objid, v_obj.schema_name, v_obj.object_name)"
PL/pgSQL function citus_drop_trigger() line 40 at PERFORM
WARNING:  cannot drop table raw_events_first_102009 because other objects depend on it
DETAIL:  constraint raw_events_second_user_id_fkey_102011 on table raw_events_second_102011 depends on table raw_events_first_102009
HINT:  Use DROP ... CASCADE to drop the dependent objects too.
CONTEXT:  while executing command on localhost:9701
SQL statement "SELECT master_drop_all_shards(v_obj.objid, v_obj.schema_name, v_obj.object_name)"
PL/pgSQL function citus_drop_trigger() line 40 at PERFORM
WARNING:  could not delete shard "raw_events_first_102009" on node "localhost:9701"
DETAIL:  Marking this shard placement for deletion
CONTEXT:  SQL statement "SELECT master_drop_all_shards(v_obj.objid, v_obj.schema_name, v_obj.object_name)"
PL/pgSQL function citus_drop_trigger() line 40 at PERFORM

```

We chatted with @anarazel on this. The simplest solution looks that we always cascade on the workers as well, which looks like a small fix . 

The simple fix is two lines change in the header + fixing the conflicts with `master_expire_table_cache`, which shouldn't take more than a day including the review as well.
(Just a note, we currently only pass CASCADE to the workers on `master_expire_table_cache` UDF, that is the reason of the conflict.)

There is a more comprehensive way of implementing this such that expand the drop trigger in a way that it recognizes the constraints (and/or other things such as triggers etc) and drops them as they are specified on the drop trigger. But, that seems much harder and not necessary for this case.

@sumedhpathak / @ozgune / @byucesoy  this looks like something we need to fix before 6.0. What do you think?

If we're going to fix this, @sumedhpathak could you assign someone on this?

### PR DESCRIPTION
**Pull request by onderkalaci**
This commit changes the dropping behaviour while dropping the shards
on the worker nodes. Fixes #927

### CODE REVIEW

**Comment by sumedhpathak** [comment]
@marcocitus could you take a look at this PR?

File: src/test/regress/sql/multi_create_table_constraints.sql [hunk 1]
```diff
@@ -249,3 +249,28 @@ DROP TABLE pk_on_part_col, uq_part_col, uq_two_columns CASCADE;
...
+CREATE TABLE raw_table_1 (user_id int, UNIQUE(user_id));
+SELECT create_distributed_table('raw_table_1', 'user_id');
+
+CREATE TABLE raw_table_2 (user_id int REFERENCES raw_table_1(user_id), UNIQUE(user_id));
+SELECT create_distributed_table('raw_table_2', 'user_id');
+
+-- see that the constraint exists
+\d raw_table_2
+
+DROP TABLE raw_table_1 CASCADE;
```
Comment (marcocitus) [inline src/test/regress/sql/multi_create_table_constraints.sql]: Maybe also do `DROP TABLE raw_table_1` to confirm we did not break the non-cascade behaviour.

File: src/test/regress/sql/multi_create_table_constraints.sql [hunk 2]
```diff
@@ -249,3 +249,28 @@ DROP TABLE pk_on_part_col, uq_part_col, uq_two_columns CASCADE;
...
+
+-- see that the constraint also dropped
+\d raw_table_2
+
+-- drop the table as well
+DROP TABLE raw_table_2;
+
+-- set the defaults back
+SET citus.shard_count = DEFAULT;
+SET citus.shard_replication_factor = DEFAULT;
```
Comment (marcocitus) [inline src/test/regress/sql/multi_create_table_constraints.sql]: These seem redundant.

File: src/backend/distributed/master/master_expire_table_cache.c [hunk 3]
```diff
@@ -174,16 +156,22 @@ DropShardsFromWorker(WorkerNode *workerNode, Oid relationId, List *shardInterval
...
+		{
+			ereport(ERROR, (errcode(ERRCODE_WRONG_OBJECT_TYPE),
+							errmsg("expire target is not a regular or foreign table")));
 		}
-	}
-
-	appendStringInfo(workerCommand, " CASCADE");
 
-	ExecuteRemoteCommand(workerNode->workerName, workerNode->workerPort, workerCommand);
+		ExecuteRemoteCommand(workerNode->workerName, workerNode->workerPort,
```
Comment (marcocitus) [inline src/backend/distributed/master/master_expire_table_cache.c]: Hmz, dropping all shards in one go is much more efficient than dropping them one by one and although it doesn't matter much here, it's what we would want to do in other cases as well.  What if we append cascade in the other places that use DROP_REGULAR_TABLE_COMMAND and leave this function as is?

**Comment by byucesoy** [comment]
Quck question; why we always cascade? Wouldn't cascading, only when user cascades correct behaviour? If I understand correctly, if we cascade all the time, we may execute commands instead of erroring out as we supposed to do.

**Comment by marcocitus** [comment]
@byucesoy The foreign key constraints on the master prevent this from happening. If we try to drop a distributed table which is referenced by a foreign key, then the drop table fails before every reaching the trigger that drops the shards. We cannot see whether the user actually typed CASCADE in the trigger, but if he didn't and there are foreign keys then we would not have reached the trigger. Therefore, adding cascade is safe, unless the shards have dependent objects that don't have corresponding objects on the master.

File: src/backend/distributed/master/master_expire_table_cache.c (hunk 3 above)
Comment (marcocitus) [inline src/backend/distributed/master/master_expire_table_cache.c]: Actually, maybe just build a list of shard names and then put it into the DROP_REGULAR_TABLE_COMMAND (with cascade) afterwards.

**Comment by marcocitus** [comment]
I addressed my own review comments, switched to injecting a list of shard names into the drop table template. 🚢

**Pull request merged**